revanced-patcher/src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -0,0 +1,277 @@
+        private val opcodes: Opcodes
+
+        /**
+         * The classes and proxied classes of the [Base] apk file.
+         */
+        val classes: ProxyBackedClassList
+
+        init {
+            MultiDexContainerBackedDexFile(object : MultiDexContainer<DexBackedDexFile> {
+                // Load all dex files from the apk module and create a dex entry for each of them.
+                private val entries = archive.readDexFiles()
+                    .mapValues { (name, data) -> BasicDexEntry(this, name, RawDexIO.readRawDexFile(data, 0, null)) }
+
+                override fun getDexEntryNames() = entries.keys.toList()
+                override fun getEntry(entryName: String) = entries[entryName]
+            }).let {
+                opcodes = it.opcodes
+                classes = ProxyBackedClassList(it.classes)
+            }
+        }
+
+        /**
+         * Write [classes] to the archive.
+         */
+        internal fun writeDexFiles() {
+            // Create patched dex files.
+            mutableMapOf<String, MemoryDataStore>().also {
+                val newDexFile = object : DexFile {
+                    override fun getClasses() =
+                        this@BytecodeData.classes.also(ProxyBackedClassList::applyProxies).toSet()
+                    override fun getOpcodes() = this@BytecodeData.opcodes
+                }
+
+                // Write modified dex files.
+                MultiDexIO.writeDexFile(
+                    true, -1, // Core count.
+                    it, Patcher.dexFileNamer, newDexFile, DexIO.DEFAULT_MAX_DEX_POOL_SIZE, null
+                )
+            }.forEach { (name, store) ->
+                archive.writeRaw(name, store.data)
+            }
+        }
+    }
+
+    /**
+     * Metadata about an [Apk] file.
+     *
+     * @param packageName The package name of the [Apk] file.
+     * @param packageVersion The package version of the [Apk] file.
+     */
+    data class PackageMetadata(val packageName: String?, val packageVersion: String?) {
+        internal constructor(manifestBlock: AndroidManifestBlock) : this(
+            manifestBlock.packageName,
+            manifestBlock.versionName
+        )
+    }
+
+    /**
+     * An [Apk] of type [Split].
+     *
+     * @param config The device configuration associated with this [Split], such as arm64_v8a, en or xhdpi.
+     * @see Apk
+     */
+    sealed class Split(val config: String, module: ApkModule) : Apk(module) {
+        override fun toString() = "split_config.$config.apk"
+
+        /**
+         * The split apk file which contains libraries.
+         *
+         * @see Split
+         */
+        class Library internal constructor(config: String, module: ApkModule) : Split(config, module) {
+            companion object {
+                /**
+                 * A set of all architectures supported by android.
+                 */
+                val architectures = setOf("armeabi_v7a", "arm64_v8a", "x86", "x86_64")
```

Comment:
Currently, the architectures are hard coded. What do you think of propagating the selection of the correct Apk types? This means the patcher would not automatically know the library APK architecture and instead require the library user to input the correct one. The benefit of that would be that the architecture wouldn't be hardcoded anymore.

Reply:
That just contains a set containing all architectures supported by android (not the current/target device). I don't think that list is likely to change any time soon so hardcoding probably isn't really an issue. It is used for determining if the split configuration is valid or not, so we cannot rely on the list of architectures that the current device uses.

Adding the ability for frontends to filter which apks are used might make sense though.